### ISSUE TITLE
Gradient-Based Adaptive Loss Weights

### ISSUE THREAD

**Issue by jrains98** [lululxvi/deepxde#331]
Hello,

I have been trying to get my PINN models to more accurately satisfy the BCs by adjusting the weights of each individual term in the loss function before adding them all together using the method described in (https://arxiv.org/abs/2001.04536).  It is similar to loss_weights in model.compile, but the weights are adjustable based on gradient information.  It requires the calculation of the maximum of the absolute values of the gradients of the PDE residual w.r.t. the training variables, as well as the mean of the absolute values of the BC residuals w.r.t the training variables.  I have been able to do so in _train_sgd by using tf.gradients, but I would like to be able to do so using the gradient information calculated by the optimizer before the optimizer applies the gradients (so I don't calculate the gradients more than once per epoch).

Thank you very much for the awesome package!

**Comment by harshil-patel-code** [comment]
Hi @jrains98: Can you please share your code. I would also like to try this for my problem of coupled PDEs

**Comment by lululxvi** [comment]
@jrains98 Great. It is doable to avoid the extra computation of gradients. In TensorFlow optimizers, `minimize` has two steps: compute_gradients() and apply_gradients()

**Comment by jrains98** [comment]
I apologize for the rather long delay in response, but thank you for your prompt response!

@harshil-patel-code  what kind of code would you like me to post?  I can post the specific training case I'm working on if that's what you want.

@lululxvi  that would be very good, since compute_gradients() and apply_gradients() only work for Adam and not LBFGS.  I'm also interested on what you think about the parallelization of these gradient calls to reduce computation time since they would have to be called for each loss term instead of the total loss.

**Comment by lululxvi** [comment]
@jrains98 Good idea. As the first step, we can make it work, and then think about what is the best way for parallel. By the way, now DeepXDE also supports PyTorch, and in PyTorch L-BFGS has the same usage as Adam.

**Comment by lululxvi** [comment]
No. DeepXDE (the same code) supports PyTorch. You only need to install PyTorch and specific the backend.

**Comment by ZazaCro** [comment]
Hi @lululxvi  @jrains98 @harshil-patel-code, I was just wondering whether the work on the dynamic loss weights, referenced in the first comment, continues or what is the status.

Thank you

**Comment by ZazaCro** [comment]
Dear @lululxvi, I have just a couple of questions.

- You mentioned to use opt.compute_gradients(loss, <list of variables>) to avoid calculating gradients multiple times in the code. Nevertheless, when opt.minimize(loss, global_step=global_step) is used we use a sum of all loss terms in comparison with the paper where we need to calculate gradients using the sum of equations' losses and the sum of boundary and initial conditions losses. Could you advise how to go about that? Do not we need to calculate the gradients separately using tf.gradients instead?
- You have also mentioned to use a callback for evaluating the adaptive loss weights. Could you advise how to pass the information back to the optimizer? It seems to me we need to define a placeholder in the model class and feed the value in using self.model.sess.run(self.model.adaptive_constant_bcs, feed_dict). This however requires updating feed_dict also at the other instances in the code. Would be there some other way?

Thank you for your help and a great tool,
Robin

**Comment by lululxvi** [comment]
@ZazaCro I think using TensorFlow 2.x or PyTorch backend would be easier, because in these two backends, the gradients are first computed and then back-prop is taken.  For example, here is the optimization for TensorFlow 2.x
https://github.com/lululxvi/deepxde/blob/af6190db34ebbcb00b88cbafd5ad9bc51de12de4/deepxde/model.py#L184
